proxy_presentation.php

```diff
@@ -0,0 +1,96 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Presentation proxy script.
+ *
+ * @package   mod_bigbluebuttonbn
+ * @copyright 2010 onwards, Blindside Networks Inc
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @author    Kenneth Hendricks (kennethhendricks@catalyst-au.net)
+ */
+
+require_once(dirname(__FILE__) . '/../../config.php');
+require_once($CFG->libdir . '/filelib.php');
+require_once(dirname(__FILE__) . '/locallib.php');
+
+$relativepath = get_file_argument();
+
+if (empty($relativepath)) {
+    exit();
+}
+
+$pathparts = explode('/', $relativepath);
+
+// If path is of format /presentation/{recording-id}/xxxx we need to check
+// user has access to the recording in question. Urls of this format contain
+// sensitive media.
+if (isset($pathparts[1]) && $pathparts[1] === 'presentation' && isset($pathparts[2])) {
+    $recordingid = $pathparts[2];
+    $recordings = bigbluebuttonbn_get_recordings_array_fetch_page([], [$recordingid]);
+
+    if (!isset($recordings[$recordingid])) {
+        throw new moodle_exception("invalidaccessparameter");
+    }
+
+    $meetingid = $recordings[$recordingid]['meetingID'];
+
+    $meetingidparts = explode('-', $meetingid);
+    if (!isset($meetingidparts[2])) {
+        throw new moodle_exception("invalidaccessparameter");
+    }
+
+    $bbbinstanceid = $meetingidparts[2];
+    $viewinstance = bigbluebuttonbn_view_instance_bigbluebuttonbn($bbbinstanceid);
+    require_login($viewinstance['course'], true, $viewinstance['cm']);
+} else {
+    require_login();
+}
+
+// One JS asset contains a non relative route we need to rewrite
+// to be relative in order to work with the proxy. We also filter out the
+// Content-Length Header in this case to avoid truncation
+$jstobereplaced = false;
+if ($relativepath === '/playback/presentation/2.0/lib/writing.js') {
+    $jstobereplaced = true;
+}
+
+$curl = new curl();
+
+if (!empty($_SERVER['HTTP_RANGE'])) {
+    $curl->setHeader("Range: {$_SERVER['HTTP_RANGE']}");
+}
+
+$curl->setopt([
+    'CURLOPT_CERTINFO'          => 1,
+    'CURLOPT_SSL_VERIFYPEER'    => true,
+    'CURLOPT_HEADERFUNCTION'    => function ($curl, $header) use ($jstobereplaced) {
+        if (!$jstobereplaced || substr($header, 0, 14) !== "Content-Length") {
+            header($header);
+        }
+        return strlen($header);
+    },
+    'CURLOPT_WRITEFUNCTION'     => function ($curl, $body) use ($jstobereplaced) {
+        if ($jstobereplaced) {
```

Comment:
same here with a ternary perhaps. This is quite picky

Reply:
Didn't go with a ternery, but rewote it a bit so its cleaner - one of the issues with just going with a ternery is we need to return the original body length even if we modified it.